simplekv/net/azurestore.py

```diff
@@ -128,3 +124,55 @@ def pickle_azure_store(store):
                              store.create_if_missing)
 
 copyreg.pickle(AzureBlockBlobStore, pickle_azure_store)
+
+
+class IOInterface(io.BufferedIOBase):
+    """
+    Class which provides a file-like interface to selectively read from a blob in the blob store.
+    """
+    def __init__(self, block_blob_service, container_name, key):
+        super(IOInterface, self).__init__()
+        self.block_blob_service = block_blob_service
+        self.container_name = container_name
+        self.key = key
+
+        blob = self.block_blob_service.get_blob_properties(container_name, key)
+        self.size = blob.properties.content_length
+        self.pos = 0
+
+    def tell(self):
+        """Returns he current offset as int. Always >= 0."""
+        return self.pos
+
+    def read(self, size=-1):
+        """Returns 'size' amount of bytes or less if there is no more data.
+        If no size is given all data is returned. size can be >= 0."""
+        with map_azure_exceptions(key=self.key):
+            if size < 0:
+                size = self.size - self.pos
+
+            end = min(self.pos + size - 1, self.size)
+            b = self.block_blob_service.get_blob_to_bytes(self.container_name,
+                                                          self.key,
+                                                          start_range=self.pos,
+                                                          end_range=end)  # end_range is inclusive
+            self.pos += end + 1
+            return b.content
+
+    def seek(self, offset, whence=0):
+        """Move to a new offset either relative or absolute. whence=0 is
+        absolute, whence=1 is relative, whence=2 is relative to the end.
+
+        Any relative or absolute seek operation which would result in a
+        negative position is undefined and that case can be ignored
+        in the implementation.
+
+        Any seek operation which moves the position after the stream
+        should succeed. tell() should report that position and read()
+        should return an empty bytes object."""
+        if whence == 0:
+            self.pos = offset
+        elif whence == 1:
+            self.pos += offset
```

Comment:
For `whence=1` and `whence=2` we can end up with an invalid (negative) position. To protect the invariants of this class we should prevent that from happening. This is also how  a normal file works in Python:

```
>>> f = open("foo")
>>> f.seek(10)
>>> f.tell()
10
>>> f.seek(-6, 1)
>>> f.tell()
4
>>> f.seek(-6, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IOError: [Errno 22] Invalid argument
>>> f.tell()
4
```
In particular, note that the failed `seek` has not altered the result of `tell`.